lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -3639,25 +3630,89 @@ bool AdjointGen::performSynthesis(FunctionSynthesisItem item) {
 // DifferentiationTask
 //===----------------------------------------------------------------------===//
 
+// Return the expected generic signature for autodiff associated functions given
+// a SILDifferentiableAttr. The expected generic signature is built from the
+// original generic signature and the attribute's requirements.
+static GenericSignature *
+getAutoDiffAssociatedFunctionGenericSignature(SILDifferentiableAttr *attr,
+                                              SILFunction *original) {
+  auto originalGenSig =
+      original->getLoweredFunctionType()->getGenericSignature();
+  if (!originalGenSig)
+    return nullptr;
+  GenericSignatureBuilder builder(original->getASTContext());
+  // Add original generic signature.
+  builder.addGenericSignature(originalGenSig);
+  // Add where clause requirements.
+  auto source =
+      GenericSignatureBuilder::FloatingRequirementSource::forAbstract();
+  for (auto &req : attr->getRequirements())
+    builder.addRequirement(req, source, original->getModule().getSwiftModule());
+  auto canGenericSig = std::move(builder)
+                           .computeGenericSignature(
+                               SourceLoc(), /*allowConcreteGenericParams=*/true)
+                           ->getCanonicalSignature();
+  return canGenericSig;
```

Comment:
The return type of this method should either be a CanGenericSignature, or you should not call getCanonicalSignature().

Reply:
Thanks! Done in https://github.com/apple/swift/pull/21973/commits/41716eb58db03ad360f3bafec46e16bf000050d1.